### PR TITLE
Remove disks endpoint using IntraBrokerDiskCapacityGoal

### PR DESCRIPTION
**Pull request by ilievladiulian**
This PR changes the implementation of the remove disks endpoint to set disks to 0 capacity and to use IntraBrokerDiskCapacityGoal to move all replicas from a disk with 0 capacity.

### CODE REVIEW

**Comment by amuraru** [comment]
@ilievladiulian do we also have a PR with the upstream project?

**Comment by ilievladiulian** [comment]
> @ilievladiulian do we also have a PR with the upstream project?

This PR addresses the comments in the upstream PR. Will also update the PR there.

**Comment by ilievladiulian** [comment]
Added exception in case the goal could not move all replicas from deleted disks so that the task ends with CompletedWithError status.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java [hunk 1]
```diff
@@ -244,12 +262,17 @@ public String name() {
 
   /**
    * Check whether the combined replica utilization is above the given disk capacity limits.
+   * If _shouldEmptyZeroCapacityDisks is true, the disk utilization is over limit only if it greater than 0.
```
Comment (cristianpetrache) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java:265]:
```suggestion
   * If _shouldEmptyZeroCapacityDisks is true, the disk utilization is over limit only if it is greater than 0.
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java [hunk 2]
```diff
@@ -43,6 +43,7 @@ public class IntraBrokerDiskCapacityGoal extends AbstractGoal {
   private static final Logger LOG = LoggerFactory.getLogger(IntraBrokerDiskCapacityGoal.class);
   private static final int MIN_NUM_VALID_WINDOWS = 1;
   private static final Resource RESOURCE = Resource.DISK;
+  private boolean _shouldEmptyZeroCapacityDisks = false;
```
Comment (cristianpetrache) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java:46]: Is this naming format specific to CC? Is it ok not to be thread safe?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java [hunk 3]
```diff
@@ -26,24 +29,28 @@
...
+    private static final int CONCURRENT_INTRA_BROKER_PARTITION_MOVEMENTS = 1;
+    private static final int CONCURRENT_LEADER_MOVEMENTS = 1;
+    private final Map<Integer, Set<String>> _brokerIdAndLogdirs;
 
     public RemoveDisksRunnable(KafkaCruiseControl kafkaCruiseControl,
                                OperationFuture future,
                                RemoveDisksParameters parameters,
                                String uuid) {
-        super(kafkaCruiseControl, future, parameters, parameters.dryRun(), parameters.stopOngoingExecution(), parameters.skipHardGoalCheck(),
+        super(kafkaCruiseControl, future, parameters, parameters.dryRun(), parameters.stopOngoingExecution(), false,
```
Comment (cristianpetrache) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:51]: Why is the param removal needed?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java (hunk 1 above)
Comment (ilievladiulian) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java]: Fixed typo in latest commit.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java (hunk 2 above)
Comment (ilievladiulian) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskCapacityGoal.java]: This seems to be the naming convention in CC. I changed the field to be final.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java (hunk 3 above)
Comment (ilievladiulian) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:51]: This was a parameter that was previously exposed in the API, although it was never used (and should not be needed for disk removal). There is no sense in skipping goal checking as this endpoint only uses one goal, the IntraBrokerDiskCapacityGoal.

Comment (cristianpetrache) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/RemoveDisksRunnable.java:51]: 👍

**Review by cristianpetrache — APPROVED** [review]

**Review by amuraru — APPROVED** [review]

**Pull request merged**